src/firewall/functions.py

```diff
@@ -101,6 +102,95 @@ def addr_family_str(family):
 ###############################################################################
 
 
+def _parse_check_fcn(parse_fcn):
+    def f(*a, **kw):
+        try:
+            parse_fcn(*a, **kw)
+        except ValueError:
+            return False
+        return True
+
+    return f
+
+
+def _parse_norm_fcn(parse_fcn, unparse_fcn):
+    def f(*a, **kw):
+        detail = parse_fcn(*a, **kw)
+        return unparse_fcn(*detail)
+
+    return f
+
+
+###############################################################################
+
+
+class EntryType:
+    class ParseFlags(enum.IntFlag):
+        NO_IP6_BRACKETS = 0x1
+
+    def __init__(self, name, fcn_parse, fcn_unparse):
+        self.name = name
+        self.parse = fcn_parse
+        self.unparse = fcn_unparse
+        self.norm = lambda *a, **kw: self._norm(*a, **kw)
+        self.check = lambda *a, **kw: self._check(*a, **kw)
+
+    def __repr__(self):
+        return f"EntryType({self.name})"
+
+    def _norm(self, *a, **kw):
+        detail = self.parse(*a, **kw)
+        return self.unparse(*detail)
```

Comment:
I'm not sure we need to expose `_norm()`, `_norm()` at all... Instead if should be part of object string representation, e.g. `str(EntryType(..))`. Implement the `_str()` function.

Reply:
The `EntryType` class is the type of something that can handle an entry.
For example, `EntryTypeAddr` is an instance of a `EntryType`. It's really a parser/unparser/checker/verifier for an IP address. So calling `str(EntryTypeAddr)` just can say "I am a parser for IP addresses".

On the other hand, `EntryTypeAddr.norm("0::1")` normalizes the "0::1" argument, that is the IPv6 address string to  "::1".

`str(EntryTypeAddr)` does not work for that.

... or I don't understand what you suggest.